src/main/java/org/isf/encounter/rest/EncounterController.java

```
@@ -136,4 +153,35 @@ public EncounterDTO updateEncounter(@PathVariable String code, @RequestBody Enco
 		encounterBrowserManager.saveEncounter(encounterToUpdated);
 		return encounter;
 	}
+
+	@GetMapping("/encounters/{code}/opds")
+	public List<OpdDTO> getOPDByEncounter(@PathVariable String code) throws OHServiceException {
+		Encounter encounter = encounterBrowserManager.getEncountersByCode(code);
+		if (encounter == null) {
+			throw new OHAPIException(new OHExceptionMessage("Encounter not found with code " + code), HttpStatus.NOT_FOUND);
+		}
+
+		List<Opd> opdList = opdManager.getOpdList(encounter.getPatient().getCode());
+		if (opdList == null || opdList.isEmpty()) {
+			throw new OHAPIException(new OHExceptionMessage("No opd found"), HttpStatus.NOT_FOUND);
+		}
+
+		List<Opd> opdListFound = null;
+		if (encounter.getStatus().toString().equals(EncounterStatus.CLOSE.toString())) {
+			opdListFound = opdList.stream()
+					.filter(opd -> opd.getDate().isAfter(encounter.getPerformedAt())
+							&& opd.getDate().isBefore(encounter.getClosedAt()))
+					.toList();
+		} else {
+			opdListFound = opdList.stream()
+					.filter(opd -> opd.getDate().isBefore(LocalDateTime.now()))
+					.toList();
+		}
+
+		if (opdListFound.isEmpty()) {
+			throw new OHAPIException(new OHExceptionMessage("No opd found"), HttpStatus.NOT_FOUND);
+		}
+
+		return opdMapper.map2DTOList(opdListFound);
+	}
```

You should throw not found error only if the encounter is not found.
If there is  not a method in the core to retrieve opds by date between, you should add it. It's not effiscient to retrieve the list, storing and filtering. You should filter directly from the data base.